### PR TITLE
BaseFilter::stream

### PR DESCRIPTION
**Pull request by debris**
build on top of #35

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ serde_derive = "1.0"
 hyper = { version = "0.10", optional = true }
 tokio-core = { version = "0.1", optional = true }
 tokio-uds = { version = "0.1", optional = true }
+tokio-timer = "0.1"
```
Comment (debris) [inline Cargo.toml]: it depends only on `futures` and `slab`

File: src/api/eth_filter.rs [hunk 2]
```diff
@@ -103,6 +107,20 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
   }
 
+  pub fn stream<'a>(&'a self, poll_interval: Duration) -> Box<Stream<Item = R, Error = Error> + 'a> where
+    T: Sync,
+    R: DeserializeOwned + Send + Sync + 'static,
+    T::Out: Send + 'static,
+  {
+    let result = Timer::default().interval(poll_interval)
+      .map_err(|_| Error::Unreachable)
+      .then(move |_| self.poll().map(|optional| optional.unwrap_or_else(Default::default)))
```
Comment (debris) [inline src/api/eth_filter.rs]: map `Future<Item = Some<Vec<_>>, _> -> Future<Item = Vec<_>, _>`

File: src/api/eth_filter.rs [hunk 3]
```diff
@@ -103,6 +107,20 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
+      .map(|res| res.into_iter().map(Ok).collect::<Vec<Result<_, Error>>>())
+      .map(stream::iter)
+      .flatten();
```
Comment (debris) [inline src/api/eth_filter.rs]: flatten it, so each event / block is a separate item of the stream

File: src/helpers.rs [hunk 4]
```diff
@@ -85,23 +85,24 @@ pub mod tests {
...
   use rpc;
   use {Result, Error, Transport};
 
   #[derive(Default)]
   pub struct TestTransport {
     asserted: usize,
-    requests: RefCell<Vec<(String, Vec<rpc::Value>)>>,
-    response: RefCell<VecDeque<rpc::Value>>,
+    requests: RwLock<Vec<(String, Vec<rpc::Value>)>>,
+    response: RwLock<VecDeque<rpc::Value>>,
```
Comment (debris) [inline src/helpers.rs]: When creating `BaseFilter::stream` i borrow `&BaseFilter`, which contains transport.

File: src/transports/ipc.rs [hunk 5]
```diff
@@ -255,7 +255,7 @@ impl Future for WriteStream {
           }
         },
         WriteState::Writing { ref buffer, ref mut current_pos } => {
-          try_ready!(Ok(self.write.poll_write()));
+          try_ready!(Ok(self.write.poll_write()) as Result<_, ()>);
```
Comment (debris) [inline src/transports/ipc.rs]: after importing `tokio_timer` I had to explicitly specify Result error. Don't know why...

File: src/api/eth_filter.rs [hunk 6]
```diff
@@ -103,6 +107,20 @@ impl<T: Transport, R> BaseFilter<T, R> {
     CallResult::new(self.transport.execute("eth_getFilterChanges", vec![id]))
   }
 
+  pub fn stream<'a>(&'a self, poll_interval: Duration) -> Box<Stream<Item = R, Error = Error> + 'a> where
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: Wouldn't it be better to consume the filter? (then RwLocks in tests are not needed)

Also can we avoid `Boxing` the stream? By returning concrete struct that owns the filer we should be able to do that.

Comment (debris) [inline src/api/eth_filter.rs]: Agree with consuming the filter! I will try prepare the struct, although it might not be trivial...

**Comment by debris** [comment]
I altered the pr, so the filter is consumed when stream is created. I had to slightly redesign the architecture of filters to avoid `derefs`. Hope you like it :)

File: src/api/eth_filter.rs [hunk 7]
```diff
@@ -1,108 +1,87 @@
...
+pub struct LogsFilter;
 
-  /// Installs a new block filter.
-  pub fn new_blocks_filter(&self) -> FilterResult<T, CallResult<U256, T::Out>, BlocksFilter<T>> {
-    FilterResult::new(
-      self.transport.clone(),
-      CallResult::new(self.transport.execute("eth_newFilter", vec![]))
-    )
-  }
+impl FilterInterface for LogsFilter {
```
Comment (rphmeier) [inline src/api/eth_filter.rs]: why not just "Filter"? docs on this trait would be nice too

File: src/api/eth_filter.rs [hunk 8]
```diff
@@ -114,104 +93,94 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
-  fn new(id: U256, transport: T) -> Self {
-    BlocksFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
+pub struct CreateFilter<T: Transport, F: FilterInterface> {
```
Comment (rphmeier) [inline src/api/eth_filter.rs]: docs? seems to be a future that resolves when the filter is created.

File: src/api/eth_filter.rs [hunk 9]
```diff
@@ -114,104 +93,94 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
+  type Item = BaseFilter<T, F>;
+  type Error = Error;
 
-  fn deref(&self) -> &Self::Target {
-    &self.base
+  fn poll(&mut self) -> Poll<Self::Item, Error> {
+    match self.future.poll() {
+      Ok(Async::Ready(x)) => Ok(Async::Ready(BaseFilter {
+        id: x,
+        transport: self.transport.clone(),
```
Comment (rphmeier) [inline src/api/eth_filter.rs]: ready occurs only once, so we could remove the `Clone` bound by making `transport` an `Option<T>` and using `take`.

File: src/api/eth_filter.rs [hunk 10]
```diff
@@ -114,104 +93,94 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
+      transport: transport
     }
   }
 }
 
-impl<T: Transport> Deref for PendingTransactionsFilter<T> {
-  type Target = BaseFilter<T, H256>;
+impl<T: Transport + Clone> EthFilter<T> {
+  /// Installs a new logs filter.
+  pub fn new_logs_filter(&self, filter: Filter) -> CreateFilter<T, LogsFilter> {
```
Comment (rphmeier) [inline src/api/eth_filter.rs]: what do you think of the naming convention `create_x_filter`?
or even just `logs()`? 

I'm thinking intended usage would be something like `my_web3.filter().logs()` which is a lot less repetitive.

File: src/api/eth_filter.rs (hunk 7 above)
Comment (debris) [inline src/api/eth_filter.rs]: there is already a structure called `Filter` in `src/types/logs.rs`

File: src/api/eth_filter.rs (hunk 10 above)
Comment (debris) [inline src/api/eth_filter.rs]: changed to `create_x_filter`

File: src/api/eth_filter.rs [hunk 11]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
-/// New blocks hashes filter.
-pub struct BlocksFilter<T: Transport> {
-  base: BaseFilter<T, H256>,
+/// Should be used to create new filter future
+pub fn create_filter<T: Transport, F: FilterInterface>(t: T, arg: Vec<rpc::Value>) -> CreateFilter<T, F> {
+  let future = CallResult::new(t.execute(F::constructor(), arg));
+  CreateFilter {
+    transport: Some(t),
+    interface: PhantomData,
+    future: future,
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: Can use a shorthnd here: `future,`

File: src/api/eth_filter.rs [hunk 12]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
+/// Future which resolves with new filter
+pub struct CreateFilter<T: Transport, F: FilterInterface> {
+  transport: Option<T>,
+  interface: PhantomData<F>,
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: It's not used so should be prefixed with `_`

File: src/api/eth_filter.rs [hunk 13]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
+  type Error = Error;
 
-  fn deref(&self) -> &Self::Target {
-    &self.base
+  fn poll(&mut self) -> Poll<Self::Item, Error> {
+    let id = try_ready!(self.future.poll());
+    let result = BaseFilter {
+      id: id,
+      transport: self.transport.take().expect("future polled after ready; qed"),
+      interface: PhantomData,
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: Same here, `_interface`.

File: src/api/eth_filter.rs [hunk 14]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
+{
+  type Item = BaseFilter<T, F>;
+  type Error = Error;
 
-  fn deref(&self) -> &Self::Target {
-    &self.base
+  fn poll(&mut self) -> Poll<Self::Item, Error> {
+    let id = try_ready!(self.future.poll());
+    let result = BaseFilter {
+      id: id,
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: Shorthand as well.

File: src/api/eth_filter.rs [hunk 15]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
-    BlocksFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
+/// Future which resolves with new filter
+pub struct CreateFilter<T: Transport, F: FilterInterface> {
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: why carry around the whole `FilterInterface`? I think just `F::Item` would be enough.

File: src/api/eth_filter.rs [hunk 16]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
-    PendingTransactionsFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
+impl<T: Transport + Clone> Namespace<T> for EthFilter<T> {
+  fn new(transport: T) -> Self where Self: Sized {
+    EthFilter {
+      transport: transport
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: shorthand.

File: src/api/eth_filter.rs [hunk 17]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
   }
 }
 
-impl<T: Transport> Deref for PendingTransactionsFilter<T> {
-  type Target = BaseFilter<T, H256>;
+impl<T: Transport + Clone> EthFilter<T> {
+  /// Installs a new logs filter.
+  pub fn create_logs_filter(&self, filter: Filter) -> CreateFilter<T, LogsFilter> {
+    let f = helpers::serialize(&filter);
+    create_filter(self.transport.clone(), vec![f])
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: `Transport` doesn't need to be `Clone`, just do `&self.transport` here.

File: src/api/mod.rs [hunk 18]
```diff
@@ -1,13 +1,12 @@
 //! `Web3` implementation
 
 mod eth;
-mod eth_filter;
+pub mod eth_filter;
```
Comment (tomusdrw) [inline src/api/mod.rs]: That's bad, better re-export everything for convenience. We don't really need to expose the module structure details imho.

File: src/transports/ipc.rs (hunk 5 above)
Comment (tomusdrw) [inline src/transports/ipc.rs]: There is probably an `Into` implementation there that breaks things.

File: src/api/eth_filter.rs [hunk 19]
```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
...
+/// Future which resolves with new filter
+pub struct CreateFilter<T: Transport, F: FilterInterface> {
+  transport: Option<T>,
+  interface: PhantomData<F>,
+  future: CallResult<U256, T::Out>,
 }
 
-impl<T: Transport> Deref for BlocksFilter<T> {
-  type Target = BaseFilter<T, H256>;
+impl<T, F> Future for CreateFilter<T, F> where
```
Comment (tomusdrw) [inline src/api/eth_filter.rs]: Just wondering if we really need a separate `Future` that resolves to `BaseFilter`. Can't the `BaseFilter` be stateful? Basically there is nothing you can do except for `and_then` on that filter, cause you need to wait for it to be created.

File: src/api/eth_filter.rs (hunk 12 above)
Comment (debris) [inline src/api/eth_filter.rs]: PhantomData does not have to be prefixed with `_`

**Pull request merged**